src/plugins/workspace/server/saved_objects/workspace_saved_objects_client_wrapper.ts

```diff
@@ -0,0 +1,517 @@
+      if (!options.ACLSearchParams) {
+        options.ACLSearchParams = {};
+      }
+
+      if (this.isRelatedToWorkspace(options.type)) {
+        // Find all "read" saved objects by object's ACL for default
+        options.ACLSearchParams.permissionModes = options.ACLSearchParams.permissionModes ?? [
+          WorkspacePermissionMode.Read,
+          WorkspacePermissionMode.Write,
+        ];
+        options.ACLSearchParams.principals = principals;
+      } else {
+        /**
+         * Workspace is a hidden type so that we need to
+         * initialize a new saved objects client with workspace enabled to retrieve all the workspaces with permission.
+         */
+        const permittedWorkspaceIds = (
+          await this.getWorkspaceTypeEnabledClient(wrapperOptions.request).find({
+            type: WORKSPACE_TYPE,
+            perPage: 999,
+            ACLSearchParams: {
+              principals,
+              /**
+               * The permitted workspace ids will be passed to the options.workspaces
+               * or options.ACLSearchParams.workspaces. These two were indicated the saved
+               * objects data inner specific workspaces. We use Library related permission here.
+               * For outside passed permission modes, it may contains other permissions. Add a intersection
+               * here to make sure only Library related permission modes will be used.
+               */
+              permissionModes: options.ACLSearchParams.permissionModes
+                ? intersection(options.ACLSearchParams.permissionModes, [
+                    WorkspacePermissionMode.LibraryRead,
+                    WorkspacePermissionMode.LibraryWrite,
+                  ])
+                : [WorkspacePermissionMode.LibraryRead, WorkspacePermissionMode.LibraryWrite],
+            },
+          })
+        ).saved_objects.map((item) => item.id);
+
+        if (options.workspaces) {
+          const permittedWorkspaces = options.workspaces.filter((item) =>
+            (permittedWorkspaceIds || []).includes(item)
+          );
+          if (!permittedWorkspaces.length) {
+            /**
+             * If user does not have any one workspace access
+             * deny the request
+             */
+            throw SavedObjectsErrorHelpers.decorateNotAuthorizedError(
+              new Error(
+                i18n.translate('workspace.permission.invalidate', {
+                  defaultMessage: 'Invalid workspace permission',
+                })
+              )
+            );
+          }
+
+          /**
+           * Overwrite the options.workspaces when user has access on partial workspaces.
+           */
+          options.workspaces = permittedWorkspaces;
+        } else {
+          /**
+           * Select all the docs that
+           * 1. ACL matches read / write / user passed permission OR
+           * 2. workspaces matches library_read or library_write OR
+           * 3. Advanced settings
```

Comment:
I guess `Advanced settings` is not a special case, having it specified in the comment may misleading?

Reply:
Address that in PR https://github.com/wanglam/OpenSearch-Dashboards/pull/4